content/blog/2020/how-selenium-works-2-navigation.md

```diff
@@ -0,0 +1,70 @@
+---
+
+After an interaction on the last weekend of January 2020, on a Selenium Issue where someone said “why can’t you just…” after I explained the issue I thought that I would start explaining commands in Selenium WebDriver and why we landed on the design that we have today.
+
+I will repeat this on every page of the series but a lot, an annoying amount sometimes, of thinking goes into how every little bit of Selenium works. 
+
+In this episode we are going to look into the sheer amount of work that goes into navigation. 
+
+{{< highlight python >}}
+from selenium import webdriver
+
+driver = webdriver.Firefox()
+driver.get("https://www.theautomatedtester.co.uk")
+{{< /highlight >}}
+
+We can see above... it looks easy enough right... no!
+
+Actually, it leads me to a pet peeve in interviews. If someone ever asks you to ever describe what happens in a browser when you type a URL and press enter there is a high probability that they have no real idea what goes into navigation. Anyway... back to Selenium and how we navigate.
+
+# `driver.get`
+
+Once we send the request over the [transportation layer]({{< relref "/blog/how-selenium-works-transport">}}) the real fun starts. We need to figure out where we want to go and tell the browser to start going there for us. Unfortunately this will lead to the first of the many problems for automators.
+
+## Certificates
+
+If you have ever worked in major corporate companies where technology is not the core then you will understand the pain that a development team might have to endure. Now double that pain and you're starting to get an idea of the pain that testers and automators have to deal with.
+
+One of these pains is certificates. Companies are cheap and will do self-signed certificates and other monstrosities. Especially in the early days of Selenium when there wasn't services like [Let's Encrypt](https://letsencrypt.org/). And even now, most Developers and QA teams rarely have access to change configurations on their test environments or on their CI servers. We needed to come up with a way to circumvent the certificates. (_This is one of the first reasons why Selenium is seen as a security risk_)
+
+So each of the browsers will allow the poorly configured certificates through when we automate. It's because if we didn't a lot of testers/developers would never be able to test their sites. 
+
+Now... that we go round the first problem we need to move on to getting the page to load.
+
+## Loading
+
+Once we have got round the certificates we get the page to load. Luckily we don't need to do anything more than the equivalent of 
+
+{{< highlight javascript >}}
+location = "https://www.theautomatedtester.co.uk"; 
+// or
+window.location.href = "https://www.theautomatedtester.co.uk";
+{{< /highlight >}}
+
+## Done... 
+
+When Selenium has "finished" a command it will return. So, we just need to wait for the page to be finished loading. To be honest, what does _"finished loading"_ even mean.
```

Comment:
I think it would be helpful to also highlight the problem of handling a `get` when the URL to navigate to wouldn't lead to a full navigation (eg. heading to an anchor in a page). There used to be a reasonably large chunk of logic in place to determine whether or not we'd need to wait for a page load to even _start_ before we would be in a position to wait for it to end.

Obviously, if we screwed up the calculation of whether to wait or not for the navigation to start, we'd either return too soon (flaky tests), or wait indefinitely (slow flaky tests). Neither option is good.

Reply:
Good point, I totally forgot about this